[BEAM-103][BEAM-130] Make Flink Source Parallel and Checkpointed

aljoscha opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one
---

CC: @mxm for the Flink parts
CC: @tgroh for the tests, I thought you could have something to say about that since you started the doc about better testing

I created a new `TestCountingSource` in the `runner-flink` package because I didn't want to depend on the data flow runner. Maybe we should move this to a common package.

dhalperi commented:
R: @mxm @tgroh @dhalperi

[I'll take a look too just to get a sense of what you're doing for sources ;)]

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/FlinkStreamingTransformTranslators.java:
```diff
@@ -284,8 +286,16 @@ public void flatMap(T s, Collector<WindowedValue<T>> collector) throws Exception
+              new UnboundedSourceWrapper<>(
+                  context.getPipelineOptions(),
+                  (UnboundedSource) transform.getSource(),
+                  context.getExecutionEnvironment().getParallelism());
```
Is this correct? We only have the the final parallelism available during runtime through the `RuntimeContext`.

aljoscha replied:
This is correct, however, I'm performing the split in the constructor to ensure that all sources agree on how things are split and also to keep it consistent when it comes to checkpoints. Once Till has his work on live rebalancing in this might have to be replaced with the `maximum shards` parameter.

mxm replied:
This shouldn't work unless the parallelism has been set explicitly on the source. The default value of the parallelism will be -1.

aljoscha replied:
Ahh man, all this parallelism business. Till is also very unhappy about this. Let's wait for what @dhalperi has to say about the splitting.

aljoscha replied:
I would also prefer doing it at runtime.

dhalperi replied:
Some of this may not translate directly because of the differences between Dataflow service and Flink clusters.

* We do indeed perform initial splitting of `UnboundedSource` at job submission time, aka when the user issues a job before submitting it to the service. This does not generally make us happy. As a signal for how far to split, we use the [`DataflowPipelineOptions.getMaxNumWorkers()`](https://github.com/apache/incubator-beam/blob/e3105c8e109535f801fd145b91b0c7aa93b86d1a/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/internal/CustomSources.java#L68) setting.

These splits are used for the entire lifetime of the job. (Hence the extra `*3` over max workers, so that we can rebalance hot splits, etc.)

We would like to additionally add dynamic splitting, such that a hot split could be subsplit. The APIs for this are not quite designed -- somehow you'd have to split given an `UnboundedSource` and a `CheckpointMark` and get 2 new pairs of the same.

We have not been planning to upgrade these APIs with these features, although we might if it becomes important to the community. What we'd like to do is a holistic re-evaluation of how sources (and not just sources -- also intermediate DoFns that have a high-fan-out of data!) can be checkpointed, restarted, and split off. See [BEAM-65](https://issues.apache.org/jira/browse/BEAM-65)

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/FlinkStreamingTransformTranslators.java:
```diff
@@ -257,6 +258,7 @@ public void flatMap(T value, Collector<WindowedValue<T>> out) throws Exception {
   private static class UnboundedReadSourceTranslator<T> implements FlinkStreamingPipelineTranslator.StreamTransformTranslator<Read.Unbounded<T>> {
 
     @Override
+    @SuppressWarnings("unchecked,rawtypes")
```
I would prefer to not set this at the method level. See below.

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/FlinkStreamingTransformTranslators.java:
```diff
@@ -284,8 +286,16 @@ public void flatMap(T s, Collector<WindowedValue<T>> collector) throws Exception
+          UnboundedSourceWrapper<T, UnboundedSource.CheckpointMark> sourceWrapper =
+              new UnboundedSourceWrapper<>(
+                  context.getPipelineOptions(),
+                  (UnboundedSource) transform.getSource(),
```
If you assign `transform.getSource()` and cast it properly, you can avoid the Suppress annotation above.

aljoscha replied:
Nope, it was there because `transform.getSource()` returns a `UnboundedSource<T, ?>` while I wanted to have a `UnboundedSource<T, UnboundedSource.CheckpointMark>`. For some reason the CheckpointMark generic parameter is `?` in almost all of the code. I'll change it to `?` here as well to avoid the `SuppressWarnings`.

mxm replied:
Yes, you can't avoid the casting because of unknown generics but I meant assigning `transform.getSource()` to a variable with a SupressWarnings annotation. This way you avoid declaring the whole method with SupressWarnings.

aljoscha replied:
You're right, yeah.

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -18,107 +18,230 @@
+    // get the splits early. we assume that the generated splits are stable,
+    // this is necessary so that the mapping of state to source is correct
+    // when restoring
+    splitSources = source.generateInitialSplits(parallelism, pipelineOptions);
```
Could we do that during runtime when the parallelism has been determined? We could save the parallelism in the checkpointed state and throw an exception if it is changed upon re-execution.

mxm replied:
Btw, if we assume the splits to be stable, we don't have to create them here but can set them at task execution.

aljoscha replied:
You're right, if they are stable we can do the `split` anywhere. I wasn't too sure, that's why I was playing it save. 😃 

@dhalperi Could you maybe chime in about whether the splits returned by `generateInitialSplits` are stable, i.e. whether we can do it in parallel on each Flink source operator.

dhalperi replied:
I'm trying to understand what this is doing. Here's the workflow from my POV:

* at submission time, user has specified some UnboundedSource in the graph. For example, `CountingInput.unbounded()` [which generates the non-negative integers `[0, 1, 2, ...)`].
* at submission time, we decide an upper bound on the number of splits. So, say, 11. We will then split the source that many ways, into `[0, 11, 22, ...)`, `[1, 12, 23, ...)`, ..., `[10, 21, 32, ...)`.

* at job runtime, we spin up workers and make some assignment of splits to workers. Supposing we have 5 workers and use round-robin, each worker will get 2-3 splits of the 11.
* the runner harness then asks each worker to start its sources and produce readers from `null` checkpoints.
* Potentially asynchronously and independently across workers, the system will ask a reader to checkpoint in order to commit some work. The local worker is required to persist that checkpoint into its local state.
* Suppose a worker fails. Then we bring it back up, but instead of giving it a `null` checkpoint we give it the last checkpoint we persisted. Thus we're fault tolerant.

My main point is that I don't understand the questions here. if `N` initial splits, should have `N` sources, `N` readers, and `N` checkpoints. If the user changes the number of VMs in a job, they probably are increasing it and those new VMs will just not have any splits. (Unless the system rebalances, aka sends a `<Split Source, Checkpoint>` pair to a new machine).

Does that make sense? Also happy to jump on a call for realtime convo.

aljoscha replied:
Thanks for the long write-up. This is pretty much how it works, yes. Except for some details in when the checkpointing happens and how we restore. (Checkpointing happens by pushing barriers through the topology. When restarting we always restart the whole topology from the last consistent checkpoint.)

I think we won't need a call for now, but thanks for the offer! With the knowledge that the order of the splits is persistent we can move the splitting to the parallel operators at runtime.

dhalperi replied:
Sorry @aljoscha  -- can you clarify for me again? I may have mis-answered.

I do not think that taking an arbitrary `UnboundedSource` and calling `generateInitialSplits(10)` will always produce the same split sources in the same order. (E.g., if I call into a service and ask for key ranges to use, it may change those over time).

I would expect that you persist the split sources AND the checkpoints together.

+ @dpmills

aljoscha replied:
Ah ok, I see. Thanks for clarifying. To summarize: We'll have to generate the initial splits once in the constructor, i.e. before the program is submitted to the cluster. Then, when checkpointing we save not only the `CheckpointMark` but also the `Source` that is responsible for that `CheckpointMark`/`Reader`. When restoring we can be certain that we match up the correct `CheckpointMark` and `Source` to create a `Reader` from the `CheckpointMark`.

Determining the maximum number of splits is a bit tricky, though. The parallelism with wich the job is going to run might not be determined yet at that point.

mxm commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -18,107 +18,230 @@
+      // we are restoring from a checkpoint, initialize the readers with the
+      // CheckpointMark. we assume a consistent order in the sources/readers,
+      // so that the order of the CheckpointMarks matches the order of
+      // our readers
```
Here we could run sanity checks on the state and check if the parallelism is still the same.

aljoscha replied:
Flink does not allow rescaling for now, and it won't, with this kind of state in the future. So I think it can never happen. If you want, I can still add it though. 😃

mxm replied:
Not necessary, I understand what you mean now with the comment.

mxm commented on runners/flink/runner/src/test/java/org/apache/beam/runners/flink/streaming/TestCountingSource.java:
```diff
@@ -0,0 +1,256 @@
+/*
```
This is copied from "org.apache.beam.sdk.runners.dataflow.TestCountingSource". I wonder if there is a way to avoid duplication by moving this class.

aljoscha replied:
Jip, I mentioned as as much in my initial comment. 😅

mxm replied:
Nice.

dhalperi replied:
I wonder if you could use CountingSource.forever()/CountingInput.unbounded() instead?

aljoscha replied:
I thought about that but couldn't find a way to get a vanilla `UnboundedCountingSource`. Should I add a static method to create one for testing, with a `VisibleForTesting` annotation?

dhalperi replied:
`CountingSource` returns a vanilla `UnboundedSource`. We could easily make it return the more concrete type.

aljoscha replied:
Ah, I needed it because the test requires an `UnboundedSource` that stops after emitting a certain number of elements. (But isn't a `BoundedSource`.). Maybe I can restructure the test to get around it.

mxm commented on runners/flink/runner/src/test/java/org/apache/beam/runners/flink/streaming/UnboundedSourceWrapperTest.java:
```diff
@@ -0,0 +1,320 @@
+    // elements later.
+    TestCountingSource source = new TestCountingSource(NUM_ELEMENTS);
+    UnboundedSourceWrapper<KV<Integer, Integer>, TestCountingSource.CounterMark> flinkWrapper =
+        new UnboundedSourceWrapper<>(options, source, 1);
```
How about another test which tests for parallelism > 1?

mxm replied:
Never mind, there is one below :)

mxm commented on runners/flink/runner/src/test/java/org/apache/beam/runners/flink/streaming/UnboundedSourceITCase.java:
```diff
@@ -1,208 +0,0 @@
-/*
```
Would it make sense to keep an ITCase, e.g. to test parallel execution of the sources?

mxm commented:
Great @aljoscha 👍 I think that brings us full compatibility with the Source API. I've made some minor comments but looks great all in all.

aljoscha commented:
@mxm Thanks for the review! Let's see what the experts have to say about the source splitting thing.

dhalperi commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -18,107 +18,255 @@
+        localSources);
+
+    readers = new ArrayList<>();
+    if (restoredState != null) {
```
this code is running per-flink-worker, or per-flink-job?

aljoscha replied:
Short version: yes, this is running per flink-worker.

For Flink there are multiple layers of complicated stuff here, the level at which we operate here we call them Operator, and an operator can have several parallel instances on several machines at runtime.

dhalperi commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -18,107 +18,255 @@
+    readers = new ArrayList<>();
+    if (restoredState != null) {
+      // we are restoring from a checkpoint, initialize the readers with the
+      // CheckpointMark. we assume a consistent order in the sources/readers,
```
this is a perfectly fine invariant to use. +1

dhalperi commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -18,107 +18,255 @@
+      try (ByteArrayInputStream bais = new ByteArrayInputStream(restoredState)) {
+        for (UnboundedSource<OutputT, CheckpointMarkT> source : localSources) {
+          CheckpointMarkT checkpointMark = checkpointMarkCoder.decode(bais, Coder.Context.OUTER);
+          readers.add(source.createReader(serializedOptions.getPipelineOptions(), checkpointMark));
```
the `OUTER` context here does not seem right. This implies that each checkpoint owns the entire stream. I would think you would probably use a `ListCoder<CheckpointMarkT>` and just decode that list. That way the individual marks would be delimited properly, etc.

aljoscha replied:
Ok, changing it.

aljoscha commented:
I addressed the comments.

@mxm I still have some `SuppressWarnings("unchecked")` in `UnboundedSourceWrapper`. Not sure if we can get rid of those. The problem is that that `UnboundedSource.split()` returns a list of `? extends UnboundedSource` and also that `Reader.checkpoint` returns a `CheckpointMark` while the checkpoint mark coder that I get from the source is of generic type `CheckpointMarkT`. @dhalperi are there plans to change that latter problem, or maybe some reasons for it being that way?

mxm commented:
Thanks for the update @aljoscha.

Regarding the split generation of the source. I think what we will have to do is to put a Mapper with parallelism 1 in front of the source. That mapper will generate the splits and send them downstream to the actual source. In Flink we can't determine the runtime parallelism beforehand so I think this is the way to go if we only want to generate splits once and achieve the best partitioning of the splits.

aljoscha commented:
@mxm That approach would get us into deep trouble, IMHO, because it would mean that we have sources in operators that don't behave as sources, which is problematic because of the way the Flink source interface works. I.e. that the source `run()` method never returns but pushes out data.

I pushed a new commit that instead always requires that a parallelism be set on the `FlinkPipelineOptions` with sensible defaults. There should now always be a valid parallelism available in the `UnboundedSourceWrapper` constructor.

mxm commented:
Thanks for fixing. That's a sensible approach to go with a default parallelism in the PipelineOptions!

dhalperi commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -18,107 +18,286 @@
 
-    reader = source.createReader(serializedOptions.getPipelineOptions(), null);
+      // restore the splitSources from the checkpoint to ensure consistent ordering
+      // do it using a transform because otherwise we would have to to
```
s/to to/to do/

aljoscha replied:
fixing

dhalperi commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -18,107 +18,286 @@
+
+      WindowedValue<OutputT> windowedValue =
+          WindowedValue.of(item, timestamp, GlobalWindow.INSTANCE, PaneInfo.NO_FIRING);
+      ctx.collectWithTimestamp(windowedValue, timestamp.getMillis());
```
do you need anything but L290 inside the synchronized block here?

dhalperi replied:
answer (from below) yes -- since we introspect the reader state below.

dhalperi commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -128,13 +307,53 @@ public void cancel() {
+  @Override
+  public void restoreState(byte[] bytes) throws Exception {
+    try (ByteArrayInputStream bais = new ByteArrayInputStream(bytes)) {
+      restoredState = checkpointCoder.decode(bais, Coder.Context.OUTER);
```
these two functions are very nice now!

dhalperi commented on runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -128,13 +307,53 @@ public void cancel() {
+        checkpoints.add(kv);
+      }
+
+      checkpointCoder.encode(checkpoints, baos, Coder.Context.OUTER);
```
might move the `try (...)` code down here for clarity, but good as is.

dhalperi commented on runners/flink/runner/src/test/java/org/apache/beam/runners/flink/streaming/UnboundedSourceWrapperTest.java:
```diff
@@ -0,0 +1,318 @@
+    } catch (SuccessException e) {
+      // success
+    } catch (Exception e) {
+      fail("We caught " + e);
```
delete this catch and stringify? let the JUnit framework handle the exception so we get a good stack trace?

tgroh replied:
These blocks should signal to the outside of the try-catch that it was caught, and the test should fail if it was not (the test has terminated without consuming all of the expected available input).

generally, if the success is the last block before test completion,
```
try {...} catch (SuccessException e) {
  return;
}
fail("Read terminated without producing expected number of outputs")
```
or another appropriate message is sufficient; the resume-from-checkpoint needs a bit of additional work around it

dhalperi commented:
LGTM

aljoscha commented:
@tgroh I changed the test, is this what you had in mind?

tgroh commented:
Yeah, that looks good.

aljoscha commented:
Alright, thanks all for reviewing! 😃  I'll merge it now.
